### PR TITLE
obs-ffmpeg: Compensate for invalid NVENC DTS when using b-frames

### PR DESCRIPTION
**Pull request by tt2468**
### Description
When using FFmpeg-based NVENC, with b-frames, and a non-1 framerate numerator (eg. `1001/60000` aka 59.94fps), the DTS values outputted by FFmpeg result in invalid DTS values.

Detect when using an unpatched FFmpeg build and correct the values accordingly.

### Motivation and Context
Invalid DTS values were causing issues with recordings being played back correctly. Streaming was seemingly unaffected in most cases. A patch has also been submitted upstream, but given how long it's expected to find its way into peoples' machines, we can also fix it on our end without too much hassle.

### How Has This Been Tested?
Streamed at 59.94, 2 b-frames, on Ubuntu (FFmpeg NVENC), and logged PTS/DTS values

Without this PR:
```
debug: pts: 0 | dts: -2
debug: pts: 3000 | dts: 998
debug: pts: 1000 | dts: 1998
debug: pts: 2000 | dts: 2998
debug: pts: 6000 | dts: 3998
debug: pts: 4000 | dts: 4998
debug: pts: 5000 | dts: 5998
debug: pts: 9000 | dts: 6998
debug: pts: 7000 | dts: 7998
debug: pts: 8000 | dts: 8998
debug: pts: 12000 | dts: 9998
debug: pts: 10000 | dts: 10998
debug: pts: 11000 | dts: 11998
debug: pts: 15000 | dts: 12998
debug: pts: 13000 | dts: 13998
debug: pts: 14000 | dts: 14998
debug: pts: 18000 | dts: 15998
debug: pts: 16000 | dts: 16998
debug: pts: 17000 | dts: 17998
debug: pts: 21000 | dts: 18998
```

With this PR:
```
info: pts: 0 | dts: -2002
info: pts: 3003 | dts: -1001
info: pts: 1001 | dts: 0
info: pts: 2002 | dts: 1001
info: pts: 6006 | dts: 2002
info: pts: 4004 | dts: 3003
info: pts: 5005 | dts: 4004
info: pts: 9009 | dts: 5005
info: pts: 7007 | dts: 6006
info: pts: 8008 | dts: 7007
info: pts: 12012 | dts: 8008
info: pts: 10010 | dts: 9009
info: pts: 11011 | dts: 10010
info: pts: 15015 | dts: 11011
info: pts: 13013 | dts: 12012
info: pts: 14014 | dts: 13013
info: pts: 18018 | dts: 14014
info: pts: 16016 | dts: 15015
info: pts: 17017 | dts: 16016
info: pts: 21021 | dts: 17017
info: pts: 19019 | dts: 18018
info: pts: 20020 | dts: 19019
info: pts: 24024 | dts: 20020
info: pts: 22022 | dts: 21021
info: pts: 23023 | dts: 22022
```

### Types of changes
- Bug fix (non-breaking change which fixes an issue)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

File: plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c [hunk 1]
```diff
@@ -292,6 +293,14 @@ static void on_first_packet(void *data, AVPacket *pkt, struct darray *da)
 					&enc->sei.array, &enc->sei.num);
 	}
 	da->capacity = da->num;
+
+	if (enc->ffve.context->max_b_frames != 0) {
+		int64_t expected_dts = -(enc->ffve.context->max_b_frames *
+					 enc->ffve.context->time_base.num);
+		if (pkt->dts != expected_dts) // Unpatched FFmpeg
+			enc->dts_offset =
```
Comment (derrod) [inline plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c:301]: Could you add a `LOG_WARNING` message here to indicate that the workaround is being applied?

File: plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c [hunk 2]
```diff
@@ -292,6 +293,14 @@ static void on_first_packet(void *data, AVPacket *pkt, struct darray *da)
...
+				expected_dts + enc->ffve.context->max_b_frames;
```
Comment (derrod) [inline plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c:302]: Shouldn't this be `expected_dts - pkt->dts`?

**Review by derrod — CHANGES_REQUESTED** [review]
Just a couple things, then I think we're good to merge this for the next beta.

Once https://github.com/obsproject/obs-deps/pull/177 is shipping we can`#ifdef` this for Linux only and once it has been upstreamed add an FFmpeg version gate.

**Review by derrod — APPROVED** [review]

**Review by RytoEX — APPROVED** [review]
Looks good in my tests. We can probably version gate this code later and possibly remove the code far in the future when we only support building against versions with the fix.

https://github.com/FFmpeg/FFmpeg/commit/9a245bdf5d7860b8bc5e5c21a105a075925b719a
https://github.com/FFmpeg/FFmpeg/commit/a6dc92968a325d331bb6dcf9b3b2248026cd1d6c
https://github.com/FFmpeg/FFmpeg/commit/00282752165a8d1e252c4a07b262d9384f618759
https://github.com/FFmpeg/FFmpeg/commit/f6facaf8c9e33472fc0be8ba5cd9666f7b0ff32d
https://github.com/FFmpeg/FFmpeg/commit/3f28c576e02edd02ee1de88303f11b870f5618a2

**Pull request merged**